src/Cassandra/Requests/RequestExecution.cs

```diff
@@ -1,84 +1,152 @@
         /// </summary>
-        /// <param name="useCurrentHost"></param>
-        public void Start(bool useCurrentHost = false)
+        private async Task SendToCurrentHostAsync()
         {
-            if (!useCurrentHost)
+            try
             {
-                //Get a new connection from the next host
-                _parent.GetNextConnection(_triedHosts).ContinueWith(t =>
+                // host needs to be re-validated using the load balancing policy
+                _connection = await _parent.ValidateHostAndGetConnectionAsync(_host, _triedHosts).ConfigureAwait(false);
+                if (_connection != null)
                 {
-                    if (t.Exception != null)
+                    Send(_request, HandleResponse);
+                    return;
+                }
+
+                RequestExecution.Logger.Warning("RequestHandler could not obtain a connection while attempting to retry with the current host.");
+            }
+            catch (Exception ex)
+            {
+                RequestExecution.Logger.Warning("RequestHandler received exception while attempting to retry with the current host: {0}", ex.ToString());
+            }
+
+            RetryExecution(false);
+        }
+
+        /// <summary>
+        /// Attempts to obtain a connection to the provided <paramref name="validHost"/> and send the request with it.
+        /// If no connection could be obtained for the provided host, then attempts to obtain a connection
+        /// for the next host, following the query plan.
+        /// </summary>
+        /// <param name="validHost">First host to which the method tries to obtain a connection.</param>
+        /// <returns></returns>
+        private async Task SendToNextHostAsync(ValidHost validHost)
+        {
+            try
+            {
+                IConnection connection = null;
+                while (connection == null)
+                {
+                    connection = await _parent.GetConnectionToValidHostAsync(validHost, _triedHosts).ConfigureAwait(false);
+                    if (connection == null)
                     {
-                        t.Exception.Handle(_ => true);
-                        HandleResponse(t.Exception.InnerException, null);
-                        return;
+                        validHost = _parent.GetNextValidHost(_triedHosts);
                     }
-                    _connection = t.Result;
-                    Send(_request, HandleResponse);
-                }, TaskContinuationOptions.ExecuteSynchronously);
-                return;
+                }
+                
+                _connection = connection;
+                _host = validHost.Host;
+                Send(_request, HandleResponse);
             }
-            if (_connection == null)
+            catch (Exception ex)
             {
-                throw new DriverInternalError("No current connection set");
+                _host = validHost.Host;
```

Comment:
Is this needed? It looks like it's setting the host to the last valid host (not necessary the one with connections).

Reply:
Ah I see... its to retry on the same host 🤦‍♂️  
thanks for the out of band explanation.

Reply:
This statement was throwing me off because, in the case of the sending, the `_host` was set before calling `Sending()`, so leaving `HandleResponse(ex, null)` for the case when you can't obtain a connection was weird to me.

Reply:
But I see that `GetConnectionToValidHostAsync()` already handles all the exception (except keyspace switches, which is fine), so this catch will only catch unexpected exceptions and that's OK to me.